src/code/TwoHeadArena.c

```diff
@@ -1,140 +1,135 @@
-        mask = ~1;
+        // Align 2 for multiples of 2
+        mask = ALIGN_MASK(2);
+    } else if (size >= 0x10) {
+        // Align 0x10 for allocations greater than 0x10
+        mask = ALIGN_MASK(0x10);
     } else {
-        mask = (size >= 0x10) ? ~0xF : 0;
+        //! @bug if size is 1 the computation below will give NULL. The mask for this case would be
+        //! more sensible as ~0, for no extra alignment
+        mask = 0;
     }
 
-    tha->tail = (void*)((((u32)tha->tail & mask) - size) & mask);
+    tha->tail = (void*)((((uintptr_t)tha->tail & mask) - size) & mask);
     return tha->tail;
 }
 
-void* THA_AllocEndAlign16(TwoHeadArena* tha, u32 size) {
-    u32 mask = ~0xF;
+/**
+ * Allocates to the tail end of the Two Head Arena with guaranteed 16-byte alignment.
+ */
+void* THA_AllocTailAlign16(TwoHeadArena* tha, size_t size) {
+    uintptr_t mask = ALIGN_MASK(0x10);
 
-    tha->tail = (void*)((((u32)tha->tail & mask) - size) & (u32)(u64)mask);
+    tha->tail = (void*)((((uintptr_t)tha->tail & mask) - size) & (uintptr_t)(u64)mask);
     return tha->tail;
 }
 
-void* THA_AllocEndAlign(TwoHeadArena* tha, u32 size, u32 mask) {
-    tha->tail = (void*)((((u32)tha->tail & mask) - size) & mask);
+/**
+ * Allocates to the tail end of the Two Head Arena using the provided mask to align the allocated region.
+ *
+ * @param tha   Arena to allocate to
+ * @param size  Size of the allocation
+ * @param mask  Mask to use to align the allocated region. To align to n-bytes where n is a power of 2, use the
+ *              ALIGN_MASK(n) macro
+ *
+ * @return Pointer to the start of the allocated block
+ */
+void* THA_AllocTailAlign(TwoHeadArena* tha, size_t size, uintptr_t mask) {
+    tha->tail = (void*)((((uintptr_t)tha->tail & mask) - size) & mask);
     return tha->tail;
 }
 
-s32 THA_GetSize(TwoHeadArena* tha) {
-    return (u32)tha->tail - (u32)tha->head;
+/**
+ * Gets the remaining size of the Two Head Arena
+ *
+ * @return Remaining size. A negative number indicates an overflow.
+ */
+s32 THA_GetRemaining(TwoHeadArena* tha) {
+    return (s32)((u8*)tha->tail - (u8*)tha->head);
 }
 
+/**
+ * Returns true if the Two Head Arena has overflown, false otherwise
+ */
 u32 THA_IsCrash(TwoHeadArena* tha) {
-    return THA_GetSize(tha) < 0;
+    return THA_GetRemaining(tha) < 0;
 }
 
-void THA_Init(TwoHeadArena* tha) {
-    tha->head = tha->bufp;
-    tha->tail = (void*)((u32)tha->bufp + tha->size);
+/**
+ * Resets the head and tail positions of the Two Head Arena, all prior allocations are effectively considered free
+ * as any new allocations will begin to overwrite them.
+ */
+void THA_Reset(TwoHeadArena* tha) {
+    tha->head = tha->start;
+    tha->tail = (u8*)tha->start + tha->size;
 }
 
-void THA_Ct(TwoHeadArena* tha, void* ptr, u32 size) {
-    tha->bufp = ptr;
+/**
+ * Creates a new Two Head Arena at `start` with available size `size`
+ */
+void THA_New(TwoHeadArena* tha, void* start, size_t size) {
```

Comment:
Should we call this Init ?

Reply:
I think `Init` makes sense, but also don't mind it staying `New` as well.